hal/armv7a/zynq7000/zynq.c

```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
+	if (dev > 24)
+		return -1;
+
+	_zynq_slcrUnlock();
+	*(zynq_common.slcr + slcr_aper_clk_ctrl) = (*(zynq_common.slcr + slcr_aper_clk_ctrl) & ~(1 << dev)) | (!!state << dev);
+	_zynq_slcrLock();
+
+	return 0;
+}
+
+
+static int _zynq_getAmbaClk(u32 dev, u32 *state)
+{
+	/* Check max dev position in amba register */
+	if (dev > 24)
+		return -1;
+
+	_zynq_slcrUnlock();
+	*state = (*(zynq_common.slcr + slcr_aper_clk_ctrl) >> dev) & 0x1;
+	_zynq_slcrLock();
+
+	return 0;
+}
+
+
+static int _zynq_setDevClk(int dev, char divisor0, char divisor1, char srcsel, char clkact0, char clkact1)
+{
+	u32 id = 0;
+
+	_zynq_slcrUnlock();
+	switch (dev) {
+		case pctl_ctrl_usb0_clk:
+		case pctl_ctrl_usb1_clk:
+			id = dev - pctl_ctrl_usb0_clk;
+			*(zynq_common.slcr + slcr_usb0_clk_ctrl + id) = (*(zynq_common.slcr + pctl_ctrl_usb0_clk + id) & ~0x00000070) | (srcsel & 0x7) << 4;
+			break;
+
+		case pctl_ctrl_gem0_rclk:
+		case pctl_ctrl_gem1_rclk:
+			id = dev - pctl_ctrl_gem0_rclk;
+			*(zynq_common.slcr + slcr_gem0_rclk_ctrl + id) = (*(zynq_common.slcr + pctl_ctrl_gem0_rclk + id) & ~0x00000011) | (!!clkact0) |
+				((!!srcsel) << 4);
+			break;
+
+		case pctl_ctrl_gem0_clk:
+		case pctl_ctrl_gem1_clk:
+			id = dev - pctl_ctrl_gem0_clk;
+			*(zynq_common.slcr + slcr_gem0_clk_ctrl + id) = (*(zynq_common.slcr + slcr_gem0_clk_ctrl + id) & ~0x03f03f71) | (!!clkact0) |
+				((srcsel & 0x7) << 4) | ((divisor0 & 0x3f) << 8) | ((divisor1 & 0x3f) << 20);
+			break;
+
+		case pctl_ctrl_smc_clk:
+			*(zynq_common.slcr + slcr_smc_clk_ctrl) = (*(zynq_common.slcr + slcr_smc_clk_ctrl) & ~0x00003f31) | (!!clkact0) |
+				((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8);
+			break;
+
+		case pctl_ctrl_lqspi_clk:
+			*(zynq_common.slcr + slcr_lqspi_clk_ctrl) = (*(zynq_common.slcr + slcr_lqspi_clk_ctrl) & ~0x00003f31) | (!!clkact0) |
+				((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8);
+			break;
+
+		case pctl_ctrl_sdio_clk:
+			*(zynq_common.slcr + slcr_sdio_clk_ctrl) = (*(zynq_common.slcr + slcr_sdio_clk_ctrl) & ~0x00003f33) | (!!clkact0) | ((!!clkact1) << 1) |
+				((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8);
+			break;
+
+		case pctl_ctrl_uart_clk:
+			*(zynq_common.slcr + slcr_uart_clk_ctrl) = (*(zynq_common.slcr + slcr_uart_clk_ctrl) & ~0x00003f33) | (!!clkact0) |
+				((!!clkact1) << 1) | ((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8);
+			break;
+
+		case pctl_ctrl_spi_clk:
+			*(zynq_common.slcr + slcr_spi_clk_ctrl) = (*(zynq_common.slcr + slcr_spi_clk_ctrl) & ~0x00003f33) | (!!clkact0) |
+				((!!clkact1) << 1) | ((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8);
+			break;
+
+		case pctl_ctrl_can_clk:
+			*(zynq_common.slcr + slcr_can_clk_ctrl) = (*(zynq_common.slcr + slcr_can_clk_ctrl) & ~0x03f03f33) | (!!clkact0) |
+				((!!clkact1) << 1) | ((srcsel & 0x3) << 4) | ((divisor0 & 0x3f) << 8) | ((divisor1 & 0x3f) << 20);
+			break;
+
+		default:
+			_zynq_slcrLock();
+			return -1;
+	}
+
+	_zynq_slcrLock();
+
+	return 0;
```

Comment:
More robust with one exit point
```suggestion
		default:
			err = -1;
			break;
	}

	_zynq_slcrLock();

	return err;
```

Reply:
Done